pennylane/decomposition/decomposition_graph.py

```diff
@@ -136,25 +136,42 @@ def __init__(
 
         # Tracks the node indices of various operators.
         self._original_ops_indices: set[int] = set()
-        self._target_ops_indices: set[int] = set()
         self._all_op_indices: dict[CompressedResourceOp, int] = {}
 
         # Stores the library of custom decomposition rules
-        self._fixed_decomps = fixed_decomps or {}
-        self._alt_decomps = alt_decomps or {}
+        fixed_decomps = fixed_decomps or {}
+        alt_decomps = alt_decomps or {}
+        self._fixed_decomps = {_to_name(k): v for k, v in fixed_decomps.items()}
+        self._alt_decomps = {_to_name(k): v for k, v in alt_decomps.items()}
 
         # Initializes the graph.
         self._graph = rx.PyDiGraph()
         self._visitor = None
 
         # Construct the decomposition graph
+        self._start = self._graph.add_node(None)
         self._construct_graph(operations)
 
-    def _get_decompositions(self, op_type) -> list[DecompositionRule]:
+    def _get_decompositions(self, op: CompressedResourceOp) -> list[DecompositionRule]:
         """Helper function to get a list of decomposition rules."""
-        if op_type in self._fixed_decomps:
-            return [self._fixed_decomps[op_type]]
-        return self._alt_decomps.get(op_type, []) + list_decomps(op_type)
+
+        op_name = _to_name(op)
+
+        if op_name in self._fixed_decomps:
+            return [self._fixed_decomps[op_name]]
+
+        decomps = self._alt_decomps.get(op_name, []) + list_decomps(op_name)
+
+        if issubclass(op.op_type, qml.ops.Adjoint):
+            decomps.extend(self._get_adjoint_decompositions(op))
+
+        elif issubclass(op.op_type, qml.ops.Pow):
+            decomps.extend(self._get_pow_decompositions(op))
+
+        elif op.op_type in (qml.ops.Controlled, qml.ops.ControlledOp):
```

Comment:
We do we use `issubclass` for adjoint and pow but `op_type in` for controlled? Seems like it might be simpler to use the same pattern for all three categories.

Reply:
Note that `AdjointOpObs` and `PowOpObs` will be going away in July, so we can potentially make decisions based around it being removed.

Reply:
For `Controlled`, we don't want any of the custom controlled operations like `CH`, `CNOT`, etc. to take this branch, but for `Adjoint`, we want all `Adjoint` ops including `AdjointOperation` to take this branch.